File: tools/ghidra/ExportDecompiledC.java
/* Export decompiled C for a list of function addresses.
 * args: <addr-list-file> <output-dir>
 * addr-list-file: one hex VA per line (0x-prefixed or bare).
 * Writes <outdir>/<va-hex>.c per function that decompiles.
 */
import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;

import java.io.File;
import java.io.FileWriter;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

public class ExportDecompiledC extends GhidraScript {
    @Override
    protected void run() throws Exception {
        String[] args = getScriptArgs();
        if (args.length < 2) {
            printerr("usage: ExportDecompiledC <addr-list-file> <outdir>");
            return;
        }
        List<String> lines = Files.readAllLines(Paths.get(args[0]));
        File outdir = new File(args[1]);
        outdir.mkdirs();

        DecompInterface ifc = new DecompInterface();
        ifc.openProgram(currentProgram);
        FunctionManager fm = currentProgram.getFunctionManager();

        int done = 0, failed = 0;
        for (String line : lines) {
            line = line.trim();
            if (line.isEmpty()) continue;
            long va = Long.parseLong(line.replace("0x", ""), 16);
            Address addr = currentProgram.getAddressFactory()
                    .getDefaultAddressSpace().getAddress(va);
            Function fn = fm.getFunctionAt(addr);
            if (fn == null) fn = fm.getFunctionContaining(addr);
            if (fn == null) { println(String.format("NOFUNC %x", va)); failed++; continue; }
            DecompileResults res = ifc.decompileFunction(fn, 120, monitor);
            if (res == null || !res.decompileCompleted()) {
                println(String.format("FAIL %x", va)); failed++; continue;
            }
            File out = new File(outdir, String.format("%08x.c", va));
            FileWriter w = new FileWriter(out);
            w.write(String.format("/* %s @ 0x%08x size=%d */%n",
                    fn.getName(), va, fn.getBody().getNumAddresses()));
            w.write(res.getDecompiledFunction().getC());
            w.close();
            done++;
        }
        ifc.dispose();
        println(String.format("ExportDecompiledC: %d written, %d failed", done, failed));
    }
}

File: tools/ghidra/ExportDecompiledC.py
# Export decompiled C for a list of function addresses.
# Ghidra Jython post-script.
#   args: <addr-list-file> <output-dir>
# addr-list-file: one hex VA per line (0x-prefixed or bare).
# Writes <outdir>/<addr>.c, one file per function that decompiles.

import os
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

args = getScriptArgs()
addr_file, outdir = args[0], args[1]
if not os.path.isdir(outdir):
    os.makedirs(outdir)

ifc = DecompInterface()
ifc.openProgram(currentProgram)
monitor = ConsoleTaskMonitor()
fm = currentProgram.getFunctionManager()
space = currentProgram.getAddressFactory().getDefaultAddressSpace()

done = failed = 0
for line in open(addr_file):
    line = line.strip()
    if not line:
        continue
    va = int(line, 16)
    addr = space.getAddress(va)
    fn = fm.getFunctionAt(addr)
    if fn is None:
        fn = fm.getFunctionContaining(addr)
    if fn is None:
        print("NOFUNC %x" % va)
        failed += 1
        continue
    res = ifc.decompileFunction(fn, 120, monitor)
    if not res.decompileCompleted():
        print("FAIL %x %s" % (va, res.getErrorMessage()))
        failed += 1
        continue
    out = os.path.join(outdir, "%08x.c" % va)
    f = open(out, "w")
    f.write("/* %s @ 0x%08x size=%d */\n" % (fn.getName(), va, fn.getBody().getNumAddresses()))
    f.write(res.getDecompiledFunction().getC())
    f.close()
    done += 1
print("ExportDecompiledC: %d written, %d failed -> %s" % (done, failed, outdir))
